File: src/deck-cost.ts
const BOARD_COVERAGE_SQFT = 0.833; // sq ft per board (standard 5/4x6x8)
const WASTE_FACTOR = 1.15;
const JOIST_SPACING_IN = 16;
const POST_AREA_SQFT = 64;
const FOOTING_PER_POST = 1;
const SCREWS_PER_100_SQFT = 350;

/** Material and installed prices per sq ft by deck material */
const DECK_PRICES: Record<string, { materialLow: number; materialHigh: number; installedLow: number; installedHigh: number }> = {
  pressure_treated: { materialLow: 15, materialHigh: 25, installedLow: 30, installedHigh: 60 },
  composite: { materialLow: 30, materialHigh: 60, installedLow: 50, installedHigh: 100 },
  cedar: { materialLow: 25, materialHigh: 40, installedLow: 40, installedHigh: 80 },
  redwood: { materialLow: 30, materialHigh: 50, installedLow: 50, installedHigh: 100 },
  hardwood: { materialLow: 35, materialHigh: 65, installedLow: 60, installedHigh: 120 },
};

const RAILING_PRICES = { low: 150, high: 400 }; // per linear ft
const STAIR_PRICES = { low: 600, high: 2000 };  // per step

export interface DeckCostInput {
  /** ft (imperial) or m (metric) */
  length: number;
  width: number;
  height: number;
  deckMaterial: keyof typeof DECK_PRICES;
  includeRailing: boolean;
  includeStairs: boolean;
  steps: number;
  units: 'imperial' | 'metric';
}

export interface DeckCostResult {
  areaSqFt: number;
  areaM2: number;
  boardsNeeded: number;
  joistsNeeded: number;
  postsNeeded: number;
  footingsNeeded: number;
  fastenersNeeded: number;
  railingLinearFt: number;
  totalMaterialLow: number;
  totalMaterialHigh: number;
  totalInstalledLow: number;
  totalInstalledHigh: number;
}

/** Calculate deck materials, framing, and cost range. */
export function calculateDeckCost(input: DeckCostInput): DeckCostResult {
  const { length, width, height, deckMaterial, includeRailing, includeStairs, steps, units } = input;

  const lengthFt = units === 'metric' ? length * 3.28084 : length;
  const widthFt = units === 'metric' ? width * 3.28084 : width;
  const heightFt = units === 'metric' ? height * 3.28084 : height;

  const areaSqFt = lengthFt * widthFt;
  const areaM2 = areaSqFt / 10.7639;

  const boardsNeeded = Math.ceil((areaSqFt / BOARD_COVERAGE_SQFT) * WASTE_FACTOR);
  const joistsNeeded = Math.ceil((lengthFt * 12) / JOIST_SPACING_IN) + 1;
  const heightMultiplier = heightFt > 4 ? 1.5 : 1;
  const postsNeeded = Math.max(4, Math.ceil((areaSqFt / POST_AREA_SQFT) * heightMultiplier));
  const footingsNeeded = postsNeeded * FOOTING_PER_POST;
  const fastenersNeeded = Math.ceil((areaSqFt / 100) * SCREWS_PER_100_SQFT);

  const perimeter = 2 * lengthFt + 2 * widthFt;
  const railingLinearFt = includeRailing ? Math.ceil(perimeter - widthFt) : 0;

  const prices = DECK_PRICES[deckMaterial] ?? DECK_PRICES.pressure_treated;
  const materialCostLow = areaSqFt * prices.materialLow;
  const materialCostHigh = areaSqFt * prices.materialHigh;
  const installedCostLow = areaSqFt * prices.installedLow;
  const installedCostHigh = areaSqFt * prices.installedHigh;

  const railingCostLow = railingLinearFt * RAILING_PRICES.low;
  const railingCostHigh = railingLinearFt * RAILING_PRICES.high;
  const stairCostLow = includeStairs ? steps * STAIR_PRICES.low : 0;
  const stairCostHigh = includeStairs ? steps * STAIR_PRICES.high : 0;

  const totalMaterialLow = materialCostLow + (includeRailing ? railingCostLow * 0.5 : 0) + (includeStairs ? stairCostLow * 0.4 : 0);
  const totalMaterialHigh = materialCostHigh + (includeRailing ? railingCostHigh * 0.5 : 0) + (includeStairs ? stairCostHigh * 0.4 : 0);
  const totalInstalledLow = installedCostLow + railingCostLow + stairCostLow;
  const totalInstalledHigh = installedCostHigh + railingCostHigh + stairCostHigh;

  return { areaSqFt, areaM2, boardsNeeded, joistsNeeded, postsNeeded, footingsNeeded, fastenersNeeded, railingLinearFt, totalMaterialLow, totalMaterialHigh, totalInstalledLow, totalInstalledHigh };
}

File: python/construction_calculators/deck_cost.py
"""Deck construction cost estimator."""

import math
from dataclasses import dataclass
from typing import Literal


_DECKING_COSTS: dict[str, dict[str, float]] = {
    "pressure_treated": {"low": 15.0, "high": 25.0},
    "cedar": {"low": 20.0, "high": 35.0},
    "redwood": {"low": 25.0, "high": 45.0},
    "composite": {"low": 30.0, "high": 60.0},
    "ipe": {"low": 35.0, "high": 70.0},
    "pvc": {"low": 25.0, "high": 50.0},
}


@dataclass
class DeckCostInput:
    """Input parameters for deck cost estimation.

    Imperial: dimensions in feet.
    Metric: dimensions in meters.
    """

    length: float
    """Deck length in feet (imperial) or meters (metric)."""
    width: float
    """Deck width in feet (imperial) or meters (metric)."""
    decking_type: str = "pressure_treated"
    """Decking material: 'pressure_treated', 'cedar', 'redwood', 'composite', 'ipe', 'pvc'."""
    include_railing: bool = True
    """Whether to include railing cost (~$40–$80/linear ft)."""
    include_stairs: bool = True
    """Whether to include one flight of stairs (~$700–$1,500)."""
    stories: int = 1
    """Number of stories above grade (affects footing/framing cost)."""
    units: Literal["imperial", "metric"] = "imperial"


@dataclass
class DeckCostResult:
    """Result of deck cost estimation."""

    area_sq_ft: float
    """Deck area in square feet."""
    area_sq_m: float
    """Deck area in square meters."""
    decking_cost_low: float
    labor_cost_low: float
    framing_cost_low: float
    railing_cost_low: float
    stair_cost_low: float
    total_low: float
    decking_cost_high: float
    labor_cost_high: float
    framing_cost_high: float
    railing_cost_high: float
    stair_cost_high: float
    total_high: float
    cost_per_sq_ft_low: float
    cost_per_sq_ft_high: float


def calculate_deck_cost(input: DeckCostInput) -> DeckCostResult:
    """Estimate deck construction cost including materials, framing, and labor.

    Args:
        input: Deck cost calculation parameters.

    Returns:
        DeckCostResult with itemized and total cost ranges.

    Examples:
        >>> from construction_calculators.deck_cost import DeckCostInput, calculate_deck_cost
        >>> inp = DeckCostInput(length=20, width=16, decking_type="composite", units="imperial")
        >>> result = calculate_deck_cost(inp)
        >>> print(f"Total: ${result.total_low:,.0f}–${result.total_high:,.0f}")
        Total: $15,280–$31,600
    """
    if input.units == "metric":
        length_ft = input.length * 3.28084
        width_ft = input.width * 3.28084
    else:
        length_ft = input.length
        width_ft = input.width

    area_sq_ft = length_ft * width_ft
    area_sq_m = area_sq_ft * 0.092903

    deck_range = _DECKING_COSTS.get(input.decking_type, _DECKING_COSTS["pressure_treated"])

    # Decking material (40% of total installed cost range)
    decking_cost_low = area_sq_ft * deck_range["low"] * 0.4
    decking_cost_high = area_sq_ft * deck_range["high"] * 0.4

    # Labor (35%)
    labor_cost_low = area_sq_ft * deck_range["low"] * 0.35
    labor_cost_high = area_sq_ft * deck_range["high"] * 0.35

    # Framing and footings (25%), adjusted for stories
    story_mult = 1.0 + (input.stories - 1) * 0.2
    framing_cost_low = area_sq_ft * deck_range["low"] * 0.25 * story_mult
    framing_cost_high = area_sq_ft * deck_range["high"] * 0.25 * story_mult

    perimeter = 2 * (length_ft + width_ft)
    if input.include_railing:
        railing_cost_low = perimeter * 40.0
        railing_cost_high = perimeter * 80.0
    else:
        railing_cost_low = railing_cost_high = 0.0

    if input.include_stairs:
        stair_cost_low = 700.0
        stair_cost_high = 1500.0
    else:
        stair_cost_low = stair_cost_high = 0.0

    total_low = decking_cost_low + labor_cost_low + framing_cost_low + railing_cost_low + stair_cost_low
    total_high = decking_cost_high + labor_cost_high + framing_cost_high + railing_cost_high + stair_cost_high

    return DeckCostResult(
        area_sq_ft=area_sq_ft,
        area_sq_m=area_sq_m,
        decking_cost_low=decking_cost_low,
        labor_cost_low=labor_cost_low,
        framing_cost_low=framing_cost_low,
        railing_cost_low=railing_cost_low,
        stair_cost_low=stair_cost_low,
        total_low=total_low,
        decking_cost_high=decking_cost_high,
        labor_cost_high=labor_cost_high,
        framing_cost_high=framing_cost_high,
        railing_cost_high=railing_cost_high,
        stair_cost_high=stair_cost_high,
        total_high=total_high,
        cost_per_sq_ft_low=total_low / area_sq_ft if area_sq_ft else 0.0,
        cost_per_sq_ft_high=total_high / area_sq_ft if area_sq_ft else 0.0,
    )
